packages/detectors.py

```diff
@@ -0,0 +1,256 @@
+    inside = C_vols[mask]
+    #Get RMS of the voxels inside, reflecting "insideness" of this scan
+    RMS_in = np.sqrt(np.mean((inside ** 2), axis=0))
+
+    #The closer this ratio is to 1, the worse the volume
+    RMS_ratio = RMS_out / RMS_in
+
+    """
+    The fourth part of the code uses the "bad components" to generate a new
+    "bad data set" and then puts this dataset through the outlier detector
+    """
+
+    #Create a boolean mask for the 10% worst PCs (meaning highest RMS ratio)
+    PC_bad = np.percentile(RMS_ratio, 90)
+    PC_bad_mask = RMS_ratio > PC_bad
+
+    U_bad = U[:, PC_bad_mask]
+    C_bad = C[PC_bad_mask]
+
+    #generates data set based on the bad PCs and (U and C matrices)
+    X_bad = U_bad.dot(C_bad).T.reshape((vol_shape + (n_vols,)))
+
+    # calculate outliers using iqr_detector
+    _, outliers = mah_detector(X_bad)
+
+    return X_bad, outliers
+
+def dvars_detector(data):
+    """
+    Calculate outliers based on RMS of differences between volumes
+    Each volume is subtracted from the following volume, then the RMS
+    is calculated for each volume difference
+
+    Outliers are then calculated by passing the RMS differences to the
+    iqr_detector function.
+
+    Parameters
+    ----------
+    data : 4D array
+        4D fMRI data
+
+    Returns
+    -------
+    dvars : 1D array
+        1d array of RMS differences of volumes equal in number to timepoints-1
+    outliers : 1D array
+        1d array of indices of outlying timepoints
+    """
+    # init dvars
+    dvars = []
+    # for 0 to timepoints-1
+    for i in range(data.shape[-1] - 1):
+        # get difference between volumes
+        vol_diff = data[...,i + 1] - data[..., i]
+        # calculate rms
+        dvars.append(np.sqrt(np.mean(vol_diff ** 2)))
+    # get outliers using iqr_detector
+    _, outliers = iqr_detector(dvars)
+    # return rms dvars and outliers
+    return dvars, outliers
+
+def mah_detector(data):
+    """ Detect outliers using Mahalanobis distance:
+    For each volume, the Mahalnobis distance is calculated via the following
+    formula:
+
+    D(t) = sqrt((X(t) - M).T.dot(S(t) ** -1).dot((X(t) - M))
+
+    where X(t) is a volume of data, M is the mean volume of the 4D data, and
+    S(t) is the covariance of X(t)
+
+    The resulting distances (D) are then input into the iqr_detector function,
+    and outliers are returned.
+
+    Parameters
+    ----------
+    data : 4D array
+        4D fMRI data
+
+    Returns
+    -------
+    D : 1D list
+        Mahalanobis distances for each volume
+    outliers : 1D list
+        indices of outlying volumes based on Mahalanobis distances
+    """
+    # init distance
+    D = []
+    # get mean of the 4d data
+    M = np.mean(data, axis=3).ravel()
+    # for each timepoint
+    for i in range(data.shape[-1]):
+        # get data for volume
+        X = data[...,i].ravel()
+        # get covariance of the data
+        S = X.dot(X.T)
```

Comment:
This is a scalar, and `S ** -1` is just `1 / S`.    In general, in numpy, `S ** -1` on an array works elementwise, and gives you `1/a` for each element, where `a` is the value in the matching position in `S`.  In MATLAB, if this was an array, `S ^ -1` does the matrix inverse, and `S .^-1` (note dot) does the elementwise reciprocal.